### PR TITLE
Install card collection

### PR DESCRIPTION
**Pull request by asarfara**
Related Ticket: https://digital-engagement.atlassian.net/browse/TVD-198

Related PR: https://github.com/dpc-sdp/tide_landing_page/pull/118

This feature adds refactoring of couple of fields so they can be used by both automated_cards and individual_cards.

**Added:**
- Added layout settings to be controlled through Drupal Display UI.
- Added content based fields through Drupal Form UI so they are not a part of  automated_card and can be used by individual_cards.
- Added new widget to allow usage conditional rendering of fields.

**Removed:**
- Removed content based fields from the code.
- Removed layout settings from the code.

**Remaining:**
- Custom logic to set default value for fields that has been taken outside of the automated_cards to make sure we can fill them based on what was set previously. This will allow us to deprecate fields rather than introducing a migration.

**Note:**
JSON still includes all the fields to maintain backward compatibility.

### CODE REVIEW

**Comment by anthony-malkoun** [comment]
> Remaining:
> 
> Custom logic to set default value for fields that has been taken outside of the automated_cards to make sure we can fill them based on what was set previously. This will allow us to deprecate fields rather than introducing a migration.

This is the outstanding issue for me. Otherwise I'm ok with how this has been implemented. It makes sense for our use case to move those fields out from the component into Drupal fields as they are shared by two components now. 

@sonnykt this module doesn't have a tagged/released version but you said that it is being used by Salsa sites? What happens to those sites if they tried to use this branch? Are those sites using this repo or the original one created by Salsa?

If this was a released module then, while deprecating fields is fine, and ideally, we would try to provide a pathway to users of the module for when the deprecated fields are removed and it becomes a breaking change. I would argue that even this change though is a breaking change as it stands as it would require action on the part of any site that was using this module. Who should provide that migration path? We can't lock ourselves down to forever providing backward compatibility and at times we will need to do breaking changes that are a bump of the major version.

@asarfara because of the time constraints of the TV project, I recommend continuing to use this branch as is. There will need to be some work do to bring this back into core SDP (mainly, getting it working for Solar as well as Together) and I would lean towards providing a migration and cleaning the code up.

**Comment by sonnykt** [comment]
@anthony-malkoun we are not using this repo for our site as it is still in private. We have our own repo and `dpc-sdp/tide_automated_listing` was originally a copy. Modifying the module in this repo should not affect our site until this module is included in SDP distro.

**Comment by anthony-malkoun** [comment]
This is a bit of a mess as Solar have it but are actually using a branch reference in their `composer.json` which is less than ideal. We'll mark this branch as a beta I think and as mentioned there will be a task to pull this together which will include a migration if required.

File: tide_automated_listing.install [hunk 1]
```diff
@@ -42,12 +42,334 @@ function tide_automated_listing_update_8002() {
                 $paragraph->save();
               }
             }
-          }
-          catch (\Exception $exception) {
+          } catch (\Exception $exception) {
```
Comment (sonnykt) [inline tide_automated_listing.install:45]: `catch` mút be on a new line.

File: src/Plugin/Field/FieldType/AutomatedListingConfiguration.php [hunk 2]
```diff
@@ -68,8 +68,8 @@ protected static function getContainer() {
    */
   public static function defaultStorageSettings() {
     return [
-      'index' => 'node',
-    ] + parent::defaultStorageSettings();
+        'index' => 'node',
```
Comment (sonnykt) [inline src/Plugin/Field/FieldType/AutomatedListingConfiguration.php:71]: CS issue: incorrect indentation.

File: src/Plugin/Field/FieldWidget/AutomatedListingConfigurationWidgetEnhanced.php [hunk 3]
```diff
@@ -0,0 +1,814 @@
...
+        unset($config['type']);
+      }
+
+      $value['value'] = Yaml::encode($config);
+    }
+
+    return $values;
+  }
+
+  protected function saveFilterValues($field_id, $wrapper) {
```
Comment (sonnykt) [inline src/Plugin/Field/FieldWidget/AutomatedListingConfigurationWidgetEnhanced.php:657]: CS issue: no PHPDoc.

File: src/Plugin/jsonapi/FieldEnhancer/AutomatedListingConfigurationEnhancer.php [hunk 4]
```diff
@@ -27,6 +27,13 @@ protected static function getIndexHelper() {
     return \Drupal::service('tide_automated_listing.sapi_index_helper');
   }
 
+  protected function setDefaultKeys() {
```
Comment (sonnykt) [inline src/Plugin/jsonapi/FieldEnhancer/AutomatedListingConfigurationEnhancer.php:30]: CS issue: no PHPDoc.

File: src/Plugin/Field/FieldWidget/AutomatedListingConfigurationWidgetEnhanced.php [hunk 5]
```diff
@@ -0,0 +1,814 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
+    $values = parent::massageFormValues($values, $form, $form_state);
+    foreach ($values as $delta => &$value) {
+      $config = [];
+      $config['index'] = $this->fieldDefinition->getFieldStorageDefinition()->getSetting('index');
+
+      /** Deprecated Fields Start */
```
Comment (sonnykt) [inline src/Plugin/Field/FieldWidget/AutomatedListingConfigurationWidgetEnhanced.php:541]: CS issue: C-style inline comment is discouraged within functions, and `/** */` should be only used to start a PHPDoc block.

**Review by sonnykt — CHANGES_REQUESTED** [review]

**Comment by asarfara** [comment]
Hey @sonnykt can you give this one last go :) I have fixed indentation issues.

**Pull request merged**